### ISSUE TITLE
SC2181 may be incorrect when command contains logical operators

### ISSUE THREAD

**Issue by skrenes** [koalaman/shellcheck#2188]
#### For bugs
- Rule Id (if any, e.g. SC1000): SC2181
- My shellcheck version (`shellcheck --version` or "online"): online (https://www.shellcheck.net/)
- [x] The rule's wiki page does not already cover this (e.g. https://shellcheck.net/wiki/SC2086)
- [x] I tried on https://www.shellcheck.net/ and verified that this is still a problem on the latest commit

#### For new checks and feature suggestions
- [x] https://www.shellcheck.net/ (i.e. the latest commit) currently gives no useful warnings about this
- [x] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Here's a snippet or screenshot that shows the problem:

```sh
#! /bin/bash

func () {
   eval "$1"
   if [[ $? -ne 0 ]]; then
      echo -e "\nAn error occured running: $1"
      exit 1
   fi
}

func2 () {
   if ! $1; then
      echo -e "\nAn error occured running: $1"
      exit 1
   fi
}

func "pwd && ls"  # executes both commands
func2 "pwd && ls" # only executes first command
```

#### Here's what shellcheck currently says:
```
Line 5:
   if [[ $? -ne 0 ]]; then
         ^-- SC2181: Check exit code directly with e.g. 'if mycmd;', not indirectly with $?.
```

#### Here's what I wanted or expected to see:
Shellcheck should not suggest changes that aren't equivalent. If there is a better way of accomplishing `func()`, I'd love to learn it. Perhaps this caveat with logical operators could be amended to this rule or the wiki entry.

**Comment by TinCanTech** [comment]
Try:
`   eval "$1" || echo -e "\nOR An error occured running: $1"`

**Comment by skrenes** [comment]
@TinCanTech unfortunately the output is the same as `func2` with the additional limitation of not being able to handle other more complicated function tasks (unless one keeps stringing commands with OR operators).

**Comment by TinCanTech** [comment]
@skrenes Not according to my test. My input is with regard to `shellcheck`, I am not debugging your script.
```
#! /bin/bash

func_a () {
   eval "$1" || echo -e "\nOR An error occured running: $1" 
}

func_b () {
   eval "$1"
   if [[ $? -ne 0 ]]; then
      echo -e "\nAn error occured running: $1"
      exit 1
   fi
}

func_c () {
   if ! $1; then
      echo -e "\nAn error occured running: $1"
      exit 1
   fi
}
```
```
$ shellcheck myscript
 
Line 9:
   if [[ $? -ne 0 ]]; then
         ^-- SC2181: Check exit code directly with e.g. 'if mycmd;', not indirectly with $?.

$ 
```

**Comment by skrenes** [comment]
My bad. I must have had a typo when I tested it. Just tested it again and you're right. So I guess that works, though it limits the size of the function unless you do something like this:
```
func_a () {
   eval "$1" || (( e=1 ))
   if [[ e -eq 1 ]]; then
        echo -e "\nAn error occurred running: $1"
        # more lines of code here...
   fi
}
```
This doesn't have any issues with [ShellCheck](https://www.shellcheck.net/), but it's basically equivalent to `func_c`/`func2` and a whole lot less readable/conventional.

**Comment by TinCanTech** [comment]
Now you simply do what `shellcheck` advised in the first place..
```
^-- SC2181: Check exit code directly with e.g. 'if mycmd;', not indirectly with $?.
```
EG:
```
func_b () {
   if eval "$1" ; then
        echo -e "\nAn error occurred running: $1"
        # more lines of code here...
   fi
}
```

**Comment by skrenes** [comment]
Oh that's right! Or to be functionally equivalent:
```
func_d () {
   if ! eval "$1"; then
        echo -e "\nAn error occurred running: $1"
        # more lines of code here...
   fi
}
```

Thanks! So then maybe the SC2181 message should read `SC2181: Check exit code directly with e.g. 'if eval $(mycmd);', not indirectly with $?.` since that is the actual equivalence.

**Comment by TinCanTech** [comment]
Your use of `eval` is part of `mycmd` not part of the `SC2181` directive.
I don't see the need for a sub-shell.
You don't have to do what `shellcheck` says, if you prefer ..

I think `shellcheck` is generally correct in its analysis and advice, in this case.

**Comment by skrenes** [comment]
Well, the trouble is, in this case, I did the former (using `eval`) and `spellcheck` suggested something that appeared to be equivalent on the surface and clearly isn't. That's what I'm reporting. Maybe the wording can be changed or a caveat added to not lead people to use a command that is not actually equivalent. So maybe something like `SC2181: Check exit code directly with e.g. 'if mycmd;' or 'if eval $mycmd;', not indirectly with $?.`

**Comment by TinCanTech** [comment]
You have two non-equivalent commands:
1. `eval` `$1`
2. `if` `!` `$1`

`shellcheck` is correct, your code is not.

```
#!/bin/bash

# https://github.com/koalaman/shellcheck/issues/2188

func_Orig () {
   eval "$1"
   if [[ $? -ne 0 ]]; then
      echo -e "\nAn error occured running: $1"
      exit 1
   else
          echo
   fi
}

func_Equiv () {
   if ! eval "$1"; then
      echo -e "\nAn error occured running: $1"
      exit 1
   else
     echo
   fi
}

func_notEq () {
   if ! $1; then
      echo -e "\nAn error occured running: $1"
      exit 1
   else
     echo
   fi
}

func_Orig   "echo '*func_Orig:' && pwd && ls"
func_Equiv  "echo '*func_Equiv:' && pwd && ls"
func_notEq  "echo '*func_notEq:' && pwd && ls"
```

**Comment by skrenes** [comment]
Oh okay, I see what you're saying. And I agree that `shellcheck` is saying to change `func_Orig` to `func_Equiv` not to `func_notEq`. That was just my incorrect interpretation. Thank you @TinCanTech for taking so much time to clarify things for me. I **really** appreciate it! Hopefully others may also benefit from this through a Google search.
